action/main.py

```diff
@@ -0,0 +1,52 @@
+import os
+import shlex
+import sys
+from pathlib import Path
+from subprocess import run, PIPE, STDOUT
+
+ACTION_PATH = Path(os.environ["GITHUB_ACTION_PATH"])
+ENV_PATH = ACTION_PATH / ".black-env"
+ENV_BIN = ENV_PATH / ("Scripts" if sys.platform == "win32" else "bin")
+OPTIONS = os.getenv("INPUT_OPTIONS", default="")
+SRC = os.getenv("INPUT_SRC", default="")
+BLACK_ARGS = os.getenv("INPUT_BLACK_ARGS", default="")
+VERSION = os.getenv("INPUT_VERSION", default="")
+
+# TODO: Uncomment these once https://github.com/actions/runner/issues/664 is resolved,
+# right now these cause more confusion than clarity :(
+# print("::group:: Setup & Install Black")
+
+run([sys.executable, "-m", "venv", str(ENV_PATH)], check=True)
+# print(f"Created virtual environment at `{ENV_PATH!s}`.")
+
+req = "black[colorama,python2]"
+if VERSION:
+    req += f"=={VERSION}"
+# TODO: remove output capturing / hiding once the log grouping works (because the logs
+# aren't out of order)
+pip_proc = run(
+    [str(ENV_BIN / "python"), "-m", "pip", "install", req],
+    stdout=PIPE,
+    stderr=STDOUT,
+    encoding="utf-8",
+)
+if pip_proc.returncode:
+    print(pip_proc.stdout)
+    print("::error::Failed to install Black.")
+    sys.exit(pip_proc.returncode)
+
+# print("::endgroup::")
+# print("::group:: Run Black")
+
+base_cmd = [str(ENV_BIN / "black")]
```

Comment:
How does windows not need `.exe` here? It can work without these days? I remember for Windows I've had to add the .exe before ... Just ignore if you're confident it works:
https://github.com/facebookincubator/ptr/blob/master/ptr.py#L545

Reply:
Definitely confident this works: https://github.com/ichard26/testblackaction/runs/2702059836?check_suite_focus=true.